<style  type="text/css"> 
</style>

<b><center>
<span style="font-size: 24pt; line-height: 1.2">
COMS W4111-002 (Spring 2021)<br>Introduction to Databases
</span>
</center></b>
</span><br>
<p>
<i><center>
<span style="font-size: 20pt; line-height: 1.2">
Homework 1: Programming - 10 Points
</span>
</center></i>

__Note:__ Please replace the information below with your last name, first name and UNI.<br><br>



<i>
<span style="font-size: 20pt; line-height: 1.2"; >
Kim_Kyungjin, kjk2170, dff9
</span>
</i>

## Introduction

### Objectives
This homework has you practice and build skill with:

- PART A: (1 point) Understanding relational databases
- PART B: (1 point) Understanding relational algebra
- PART C: (1 point) Cleaning data
- PART D: (1 point) Performing simple SQL queries to analyze the data.
- PART E: (6 points) CSVDataTable.py


__Note:__ The motivation for PART E may not be clear. The motivation will become clearer as the semester proceeds. The purpose of PART E is to get you started on programming in Python and manipulating data.



### Submission
1. File > Download as > PDF via latex (.PDF) (Use the "File" menu option on the Jupyter notebook tool bar, not the option on the browser tool bar).
2. Upload .pdf and .ipynb to GradeScope
3. Upload CSVDataTable.py and CSVDataTable_Tests.py

**This assignment is due January 29, 11:59 pm EDT**


### Collaboration
- You may use any information you get in TA or Prof. Ferguson's office hours, from lectures or from recitations.
- You may use information that you find on the web.
- You are NOT allowed to collaborate with other students outside of office hours.


# Part A: Written

1. What is a database management system?

    <i> collection of interrelated data(usually referred as database) and a set of programs to access those data. </i>

2. What is a foreign key?

    <i> When attributes A of relation r1 to the unique key B of relation r2 states that on any database instance, if value of A for each tuple in r1 has the same value of B for some tuple in r2, A is the "FOREIGN KEY" from r1 referencing r2. Furthermore, r1 is called the referencing relation of the foreign-key constraint, and r2 is called the referenced relation.  </i>

3. What is a primary key?

    <i> candidate key that is chosen as the principal means of identifying tuples within a relation. Note that candidate key is the superkeys which no proper subset is a superkey. </i>

4. What are 4 different types of DBMS relationships, give a brief explanaition for each?

    <i> One-to-One(entity in A is associated with at most one entity in B, and vice versa) 
    One-to-many(entity in A is associated with any number of entities in B. Entity in B can be assciated with at most one entity in A) 
    Many-to-one(An entity in A is associated with at most one entity in B. An entity in B can be associated with any number of entities in A) 
    Many-to-many(An entity in A is associated with any number (zero or more) of entities in B, and an entity in B is associated with any number (zero or more) ofentities in A). </i>

5. What is an ER model? 

    <i> E-R(Entity-Relationship) data model: Model to facilitate database desgin by allowing specification of an enterpirse schema that represents the overall logical structure of a database. Where entity is a object in the real world that is distinguishable from all other objects and relationship is an associataion among several entities.  </i>


6. Using Lucidchart draw an example of a logical ER model using Crow's Foot notation for Columbia classes. The entity types are:
    - Students, Professors, and Classes.
    - The relationships are:
        - A Class has exactly one Professor.
        - A Student has exactly one professor who is an _advisor._
        - A Professor may advise 0, 1 or many Students.
        - A Class has 0, 1 or many enrolled students.
        - A Student enrolls in 0, 1 or many Classes.
        
        
- You can define what you think are common attributes for each of the entity types. Do not define more than 5 or 6 attributes per entity type.


- In this example, explicitly show an example of a primary-key, foreign key, one-to-many relationship, and many-to-many relationship.


__Notes:__
- If you have not already done so, please register for a free account at Lucidchart.com. You can choose the option at the bottom of the left pane to add the ER diagram shapes.
- You can take a screen capture of you diagram and save in the zip directory that that contains you Jupyter notebook. Edit the following cell and replace "Boromir.jpg" with the name of the file containing your screenshot.
    
    

   <i> Use the following line to upload a photo of your Luicdchart. <i/>
   <img src="Boromir.jpg">

# Part B: Relational Algebra

You will use [the online relational calculator](https://dbis-uibk.github.io/relax/landing), choose the “Karlsruhe University of Applied Sciences” dataset.

An anti-join is a form of join with reverse logic. Instead of returning rows when there is a match (according to the join predicate) between the left and right side, an anti-join returns those rows from the left side of the predicate for which there is no match on the right. 

The Anti-Join Symbol is ▷.

Consider the following relational algebra expression and result.

/* (1) Set X = The set of classrooms in buildings Taylor or Watson. */

        X = σ building='Watson' ∨ building='Taylor' (classroom)

/* (2) Set Y =  The Anti-Join of department and X */

        Y = (department ▷ X)

/* (3) Display the rows in Y. */

        Y
        
   <img src="ra.png">


1. Find an alternate expression to (2) that computes the correct answer given X. Display the execution of your query below.


    <i>department⋉(σ building ≠ 'Watson' and building ≠ 'Taylor' (classroom)) </i>
<img src = "Relational Algebra.png">

# Part C: Data Clean Up

## Please note: You MUST make a new schema using the lahmansdb_to_clean.sql file provided in the data folder. 

Use thelahmansdb_to_clean.sql file to make a new schema containing the raw data. The lahman database you created in Homework 0 has already been cleaned with all the constraints and will be used for Part D. Knowing how to clean data and add integrity constraints is very important which is why you go through the steps in part C. 

TLDR: If you use the HW0 lahman schema for this part you will get a lot of errors and recieve a lot of deductions. 

In [14]:
# You will need to follow instructions from HW 0 to make a new schema, import the data. 
# Connect to the unclean schema below by setting the database host, user ID and password.
%load_ext sql
%sql mysql+pymysql://admin:Rlarudwls45!@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean

The sql extension is already loaded. To reload it, use:
  %reload_ext sql



Data cleanup: For each table we want you to clean, we have provided a list of changes you have to make. You can reference the cleaned lahman db for inspiration and guidance, but know that there are different ways to clean the data and you will be graded for your choice rationalization. You should make these changes through DataGrip's workbench's table editor and/or using SQL queries. In this part you will clean two tables: People and Batting. 


### You must have: 
- A brief explanation of why you think the change we requested is important.
- What change you made to the table.
- Any queries you used to make the changes, either the ones you wrote or the Alter statements provided by DataGrip's editor. 
- Executed the test statements we provided
- The cleaned table's new create statement (after you finish all the changes) 


### Overview of Changes:

People Table 

   0. Primary Key (Explanation is given, but you still must add the key to your table yourself)
   1. Empty strings to NULLs
   2. Column typing
   3. isDead column
   4. deathDate and birthDate column

Batting Table

   1. Empty strings to NULLs
   2. Column typing
   3. Primary Key
   4. Foreign Key
   

   
### How to make the changes: 

__Using the Table Editor:__

When you hit apply, a popup will open displaying the ALTER statments sql generates. Copy the sql provided first and paste it into this notebook. Then you can apply the changes. This means that you are NOT executing the ALTER statements through your notebook. 


1. Right click on the table > Modify Table...
<img src="modify.png" width="400" height="800">

2. Keys > press the + button > input the parameters > Execute 
    OR Keys > press the + button > input the parameters > copy and paste the script generated under "SQL Script" and paste into your notebook > Run the cell in jupyter notebook
<img src="pk.png" width="600" height="1200">




__Using sql queries:__

Copy paste any queries that you write manually into the notebook as well!


<hr style="height:2px">

## People Table

### 0) EXAMPLE: Add a Primary Key 

(Solutions are given but make sure you still do this step in workbench!)

#### Explanation

We want to add a Primary Key because we want to be able to uniquely identify rows within our data. A primary key is also an index, which allows us to locate data faster. 

#### Change

I added a Primary Key on the playerID column and made the datatype VARCHAR(15)

__Note:__ This is for demonstration purposes only. playerID __is not__ a primary key for fielding.

#### SQL

~~~~sql
ALTER TABLE `lahmansdb_to_clean`.`people`
CHANGE COLUMN `playerID` `playerID` VARCHAR(15) NOT NULL ,
ADD PRIMARY KEY (`playerID`);
~~~~

#### Tests

In [15]:
%sql SHOW KEYS FROM people WHERE Key_name = 'PRIMARY'
    

   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
1 rows affected.


[('people', 0, 'PRIMARY', 1, 'playerID', 'A', 19512, None, None, '', 'BTREE', '', '', 'YES', None)]

### 1) Convert all empty strings to NULL

#### Explanation

_As a start of data clean up, we want to change empty string to NULL 
to represent missing information/inapplicable information _

#### Change

_Select each attributes in people , set empty string to NULL_

#### SQL

~~~~sql
UPDATE people SET playerID = NULL WHERE playerID = '';
UPDATE people SET birthYear = NULL WHERE birthYear = '';
UPDATE people SET birthMonth = NULL WHERE birthMonth = '';
UPDATE people SET birthDay = NULL WHERE birthDay = '';
UPDATE people SET birthCountry = NULL WHERE birthCountry = '';
UPDATE people SET birthState = NULL WHERE birthState = '';
UPDATE people SET birthCity = NULL WHERE birthCity = '';
UPDATE people SET deathYear = NULL WHERE deathYear = '';
UPDATE people SET deathMonth = NULL WHERE deathMonth = '';
UPDATE people SET deathDay = NULL WHERE deathDay = '';
UPDATE people SET deathCountry = NULL WHERE deathCountry = '';
UPDATE people SET deathState = NULL WHERE deathState = '';
UPDATE people SET deathCity = NULL WHERE deathCity = '';
UPDATE people SET nameFirst = NULL WHERE nameFirst = '';
UPDATE people SET nameLast = NULL WHERE nameLast = '';
UPDATE people SET nameGiven = NULL WHERE nameGiven = '';
UPDATE people SET weight = NULL WHERE weight = '';
UPDATE people SET height = NULL WHERE height = '';
UPDATE people SET bats = NULL WHERE bats = '';
UPDATE people SET throws = NULL WHERE throws = '';
UPDATE people SET debut = NULL WHERE debut = '';
UPDATE people SET finalGame = NULL WHERE finalGame = '';
UPDATE people SET retroID = NULL WHERE retroID = '';
UPDATE people SET bbrefID = NULL WHERE bbrefID = '';
~~~~

#### Tests

In [16]:
%sql SELECT * FROM people WHERE birthState is NULL LIMIT 10;

   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
10 rows affected.


[('agostju01', '1958', '2', '23', datetime.date(1958, 2, 23), None, 'P.R.', None, 'Rio Piedras', 'N', None, None, None, None, None, None, 'Juan', 'Agosto', 'Juan Roberto', 190, 74, 'L', 'L', datetime.date(1981, 9, 7), datetime.date(1993, 6, 19), 'agosj001', 'agostju01'),
 ('aguaylu01', '1959', '3', '13', datetime.date(1959, 3, 13), None, 'P.R.', None, 'Vega Baja', 'N', None, None, None, None, None, None, 'Luis', 'Aguayo', 'Luis', 173, 69, 'R', 'R', datetime.date(1980, 4, 19), datetime.date(1989, 9, 30), 'agual001', 'aguaylu01'),
 ('ainsmed01', '1890', '2', '4', datetime.date(1890, 2, 4), datetime.date(1981, 9, 6), 'Russia', None, None, 'Y', '1981', '9', '6', 'USA', 'FL', 'Fort Lauderdale', 'Eddie', 'Ainsmith', 'Edward Wilbur', 180, 71, 'R', 'R', datetime.date(1910, 8, 9), datetime.date(1924, 7, 21), 'ainse101', 'ainsmed01'),
 ('albaljo01', '1982', '10', '30', datetime.date(1982, 10, 30), None, 'P.R.', None, 'San Juan', 'N', None, None, None, None, None, None, 'Jonathan', 'Albaladejo', 'Jonathan', 270, 77, 'R', 'R', datetime.date(2007, 9, 5), datetime.date(2012, 7, 20), 'albaj001', 'albaljo01'),
 ('alberjo01', '1969', '6', '29', datetime.date(1969, 6, 29), None, 'P.R.', None, 'San Juan', 'N', None, None, None, None, None, None, 'Jose', 'Alberro', 'Jose Edgardo', 190, 74, 'R', 'R', datetime.date(1995, 4, 27), datetime.date(1997, 8, 5), 'albej001', 'alberjo01'),
 ('albieoz01', '1997', '1', '7', datetime.date(1997, 1, 7), None, 'Curacao', None, 'Willemstad', 'N', None, None, None, None, None, None, 'Ozzie', 'Albies', 'Ozhaino Jurdy Jiandro', 165, 68, 'B', 'R', datetime.date(2017, 8, 1), datetime.date(2020, 9, 27), 'albio001', 'albieoz01'),
 ('alcarlu01', '1941', '6', '20', datetime.date(1941, 6, 20), None, 'P.R.', None, 'Humacao', 'N', None, None, None, None, None, None, 'Luis', 'Alcaraz', 'Angel Luis', 165, 69, 'R', 'R', datetime.date(1967, 9, 13), datetime.date(1970, 6, 7), 'alcal101', 'alcarlu01'),
 ('alicelu01', '1965', '7', '29', datetime.date(1965, 7, 29), None, 'P.R.', None, 'Santurce', 'N', None, None, None, None, None, None, 'Luis', 'Alicea', 'Luis Rene', 165, 69, 'B', 'R', datetime.date(1988, 4, 23), datetime.date(2002, 9, 25), 'alicl001', 'alicelu01'),
 ('alomaro01', '1968', '2', '5', datetime.date(1968, 2, 5), None, 'P.R.', None, 'Ponce', 'N', None, None, None, None, None, None, 'Roberto', 'Alomar', 'Roberto', 184, 72, 'B', 'R', datetime.date(1988, 4, 22), datetime.date(2004, 9, 5), 'alomr001', 'alomaro01'),
 ('alomasa01', '1943', '10', '19', datetime.date(1943, 10, 19), None, 'P.R.', None, 'Salinas', 'N', None, None, None, None, None, None, 'Sandy', 'Alomar', 'Santos', 140, 69, 'B', 'R', datetime.date(1964, 9, 15), datetime.date(1978, 9, 30), 'aloms101', 'alomasa01')]

### 2) Change column datatypes to appropriate values (ENUM, INT, VARCHAR, DATETIME, ETC)

#### Explanation

_Setting up column datatype is important to understand the type of data stored and how they interact each other. Each column is examined and appropriate data type is selected. _

#### Change

_Alter people table then modify column with appropirate data type. Note that maximum length of each column should be carefully inspected before setting up length for the datatype(For example, nameGinven Column has maximum length of 43)_

#### SQL

~~~~sql
ALTER table people modify COLUMN playerID varchar(15);
ALTER table people modify COLUMN birthYear VARCHAR(4);
ALTER table people modify COLUMN birthMonth VARCHAR(2);
ALTER table people modify COLUMN birthDay VARCHAR(2);
ALTER table people modify COLUMN birthCountry varchar(15);
ALTER table people modify COLUMN birthState varchar(25);
ALTER table people modify COLUMN birthCity varchar(30);
ALTER table people modify COLUMN deathYear VARCHAR(4);
ALTER table people modify COLUMN deathMonth VARCHAR(2);
ALTER table people modify COLUMN deathDay VARCHAR(2);
ALTER table people modify COLUMN deathCountry varchar(15);
ALTER table people modify COLUMN deathState varchar(25);
ALTER table people modify COLUMN deathCity varchar(30);
ALTER table people modify COLUMN nameFirst varchar(15);
ALTER table people modify COLUMN nameLast varchar(15);
ALTER table people modify COLUMN nameGiven varchar(45);
ALTER table people modify COLUMN weight INT(4);
ALTER table people modify COLUMN height INT(2);
ALTER table people modify COLUMN bats CHAR(1);
ALTER table people modify COLUMN throws ENUM('R', 'L', 'S');
ALTER table people modify COLUMN debut DATE;
ALTER table people modify COLUMN finalGame DATE;
ALTER table people modify COLUMN retroID VARCHAR(15);
ALTER table people modify COLUMN bbrefID VARCHAR(15);

~~~~
to inspect maximum value in each column,

~~~~sql
SELECT *
FROM people
ORDER BY LENGTH(columnname) DESC
LIMIT 1;
~~~~

### 3) Add an isDead Column that is either 'Y' or 'N'
- Some things to think of: What data type should this column be? How do you know if the player is dead or not? Maybe you do not know if the player is dead.
- You do not need to make guesses about life spans, etc. Just apply a simple rule.

'Y' means the player is dead

'N' means the player is alive

#### Explanation

_isDead column is added in people schema to identify whether player is dead or not. If deathYear column is not empty, then player is regarded as dead _

#### Change

_First add column with VARCHAR(1) since only Y or N is accepted. Then update column value based on deathYear column value_

#### SQL

~~~~sql
ALTER table people
ADD COLUMN isDead VARCHAR(1) AFTER birthCity;

UPDATE people SET isDead = 'N' WHERE deathYear IS NULL;
UPDATE people SET isDead = 'Y' WHERE deathYear IS NOT NULL;
~~~~

#### Tests

In [17]:
%sql SELECT * FROM people WHERE isDead = 'N' limit 10;

   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
10 rows affected.


[('aardsda01', '1981', '12', '27', datetime.date(1981, 12, 27), None, 'USA', 'CO', 'Denver', 'N', None, None, None, None, None, None, 'David', 'Aardsma', 'David Allan', 215, 75, 'R', 'R', datetime.date(2004, 4, 6), datetime.date(2015, 8, 23), 'aardd001', 'aardsda01'),
 ('aaronha01', '1934', '2', '5', datetime.date(1934, 2, 5), None, 'USA', 'AL', 'Mobile', 'N', None, None, None, None, None, None, 'Hank', 'Aaron', 'Henry Louis', 180, 72, 'R', 'R', datetime.date(1954, 4, 13), datetime.date(1976, 10, 3), 'aaroh101', 'aaronha01'),
 ('aasedo01', '1954', '9', '8', datetime.date(1954, 9, 8), None, 'USA', 'CA', 'Orange', 'N', None, None, None, None, None, None, 'Don', 'Aase', 'Donald William', 190, 75, 'R', 'R', datetime.date(1977, 7, 26), datetime.date(1990, 10, 3), 'aased001', 'aasedo01'),
 ('abadan01', '1972', '8', '25', datetime.date(1972, 8, 25), None, 'USA', 'FL', 'Palm Beach', 'N', None, None, None, None, None, None, 'Andy', 'Abad', 'Fausto Andres', 184, 73, 'L', 'L', datetime.date(2001, 9, 10), datetime.date(2006, 4, 13), 'abada001', 'abadan01'),
 ('abadfe01', '1985', '12', '17', datetime.date(1985, 12, 17), None, 'D.R.', 'La Romana', 'La Romana', 'N', None, None, None, None, None, None, 'Fernando', 'Abad', 'Fernando Antonio', 235, 74, 'L', 'L', datetime.date(2010, 7, 28), datetime.date(2019, 9, 28), 'abadf001', 'abadfe01'),
 ('abbotgl01', '1951', '2', '16', datetime.date(1951, 2, 16), None, 'USA', 'AR', 'Little Rock', 'N', None, None, None, None, None, None, 'Glenn', 'Abbott', 'William Glenn', 200, 78, 'R', 'R', datetime.date(1973, 7, 29), datetime.date(1984, 8, 8), 'abbog001', 'abbotgl01'),
 ('abbotje01', '1972', '8', '17', datetime.date(1972, 8, 17), None, 'USA', 'GA', 'Atlanta', 'N', None, None, None, None, None, None, 'Jeff', 'Abbott', 'Jeffrey William', 190, 74, 'R', 'L', datetime.date(1997, 6, 10), datetime.date(2001, 9, 29), 'abboj002', 'abbotje01'),
 ('abbotji01', '1967', '9', '19', datetime.date(1967, 9, 19), None, 'USA', 'MI', 'Flint', 'N', None, None, None, None, None, None, 'Jim', 'Abbott', 'James Anthony', 200, 75, 'L', 'L', datetime.date(1989, 4, 8), datetime.date(1999, 7, 21), 'abboj001', 'abbotji01'),
 ('abbotku01', '1969', '6', '2', datetime.date(1969, 6, 2), None, 'USA', 'OH', 'Zanesville', 'N', None, None, None, None, None, None, 'Kurt', 'Abbott', 'Kurt Thomas', 180, 71, 'R', 'R', datetime.date(1993, 9, 7), datetime.date(2001, 4, 13), 'abbok002', 'abbotku01'),
 ('abbotky01', '1968', '2', '18', datetime.date(1968, 2, 18), None, 'USA', 'MA', 'Newburyport', 'N', None, None, None, None, None, None, 'Kyle', 'Abbott', 'Lawrence Kyle', 200, 76, 'L', 'L', datetime.date(1991, 9, 10), datetime.date(1996, 8, 24), 'abbok001', 'abbotky01')]

### 4) Add a deathDate and birthDate column
Some things to think of: What do you do if you are missing information? What datatype should this column be?

You have to create this column from other columns in the table.

#### Explanation

_deathDate and birthDate refered from birthYear/Month/Day, deathYear/Month/Day columns. Note that if at least one of column is null, then birthDate/deatDate is selected as null(unknown) since we do not know  the exact birth/death date._

#### Change

_Make two new columns name birthDate and deathDate with datatype date. concat values describe above if each value is not null_

#### SQL

~~~~sql
ALTER table people
ADD COLUMN birthDate date AFTER birthDay ;

ALTER table people
ADD COLUMN deathDate date AFTER birthDate;


update people SET birthDate = CONCAT(birthYear, '-',birthMonth,'-',birthDay)
WHERE birthYear is not null and birthMonth is not null and birthDay is not null;

update people SET deathDate = CONCAT(deathYear, '-',deathMonth,'-',deathDay)
WHERE deathYear is not null and deathMonth is not null and deathDay is not null;
~~~~

#### Tests

In [18]:
%sql SELECT deathDate FROM people WHERE deathDate >= '2005-01-01' ORDER BY deathDate ASC LIMIT 10;

   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
10 rows affected.


[(datetime.date(2005, 1, 4),),
 (datetime.date(2005, 1, 7),),
 (datetime.date(2005, 1, 9),),
 (datetime.date(2005, 1, 10),),
 (datetime.date(2005, 1, 21),),
 (datetime.date(2005, 1, 22),),
 (datetime.date(2005, 1, 31),),
 (datetime.date(2005, 2, 4),),
 (datetime.date(2005, 2, 8),),
 (datetime.date(2005, 2, 11),)]

In [19]:
%sql SELECT birthDate FROM people WHERE birthDate <= '1965-01-0' ORDER BY birthDate ASC LIMIT 10;


   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
10 rows affected.


[(datetime.date(1820, 4, 17),),
 (datetime.date(1824, 10, 5),),
 (datetime.date(1832, 9, 17),),
 (datetime.date(1832, 10, 23),),
 (datetime.date(1835, 1, 10),),
 (datetime.date(1836, 2, 29),),
 (datetime.date(1837, 12, 26),),
 (datetime.date(1838, 3, 10),),
 (datetime.date(1838, 7, 16),),
 (datetime.date(1838, 8, 27),)]

### Final CREATE Statement

To find the create statement:

- Right click on the table name in workbench
- Select 'Copy to Clipboard'
- Select 'Create Statement' 

The create statement will now be copied into your clipboard and can be pasted into the cell below. 

~~~~sql
cant find it from datagrip;
~~~~

<hr style="height:2px">

## Batting Table

### 1) Convert all empty strings to NULL

#### Explanation

_Update every value of empty string in batting table to NULL_

#### Change

_Select each attributes in batting , set empty string to NULL_

#### SQL

~~~~sql

UPDATE batting SET playerID = NULL WHERE playerID ='';
UPDATE batting SET yearID = NULL WHERE yearID ='';
UPDATE batting SET stint = NULL WHERE stint ='';
UPDATE batting SET teamID = NULL WHERE teamID ='';
UPDATE batting SET lgID = NULL WHERE lgID ='';
UPDATE batting SET G = NULL WHERE G ='';
UPDATE batting SET AB = NULL WHERE AB ='';
UPDATE batting SET R = NULL WHERE R ='';
UPDATE batting SET H = NULL WHERE H ='';
UPDATE batting SET 2B = NULL WHERE 2B ='';
UPDATE batting SET 3B = NULL WHERE 3B ='';
UPDATE batting SET HR = NULL WHERE HR ='';
UPDATE batting SET RBI = NULL WHERE RBI ='';
UPDATE batting SET SB = NULL WHERE SB ='';
UPDATE batting SET CS = NULL WHERE CS ='';
UPDATE batting SET BB = NULL WHERE BB ='';
UPDATE batting SET SO = NULL WHERE SO ='';
UPDATE batting SET IBB = NULL WHERE IBB ='';
UPDATE batting SET HBP = NULL WHERE HBP ='';
UPDATE batting SET SH = NULL WHERE SH ='';
UPDATE batting SET SF = NULL WHERE SF ='';
UPDATE batting SET GIDP = NULL WHERE GIDP ='';
~~~~

#### Tests

In [20]:
%sql SELECT count(*) FROM lahmansdb_to_clean.batting where RBI is NULL;

   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
1 rows affected.


[(756,)]

### 2) Change column datatypes to appropriate values (ENUM, INT, VARCHAR, DATETIME, ETC)

#### Explanation

_Change column data type of batting appropriatly by insepction_

#### Change

_change column data type by use alter/modify. Note that INT is selected rather than ENUM in stint due to range of values(too many)_

#### SQL

~~~~sql
ALTER table batting modify COLUMN playerID VARCHAR(15);
ALTER table batting modify COLUMN yearID CHAR(4);
ALTER table batting modify COLUMN stint INT(1);
ALTER table batting modify COLUMN teamID VARCHAR(3);
ALTER table batting modify COLUMN lgID VARCHAR(2);
ALTER table batting modify COLUMN G int(3);
ALTER table batting modify COLUMN AB int(3);
ALTER table batting modify COLUMN R int(3);
ALTER table batting modify COLUMN H int(3);
ALTER table batting modify COLUMN 2B int(3);
ALTER table batting modify COLUMN 3B int(3);
ALTER table batting modify COLUMN HR int(3);
ALTER table batting modify COLUMN RBI int(3);
ALTER table batting modify COLUMN SB int(3);
ALTER table batting modify COLUMN CS int(3);
ALTER table batting modify COLUMN BB int(3);
ALTER table batting modify COLUMN SO int(3);
ALTER table batting modify COLUMN IBB int(3);
ALTER table batting modify COLUMN HBP int(3);
ALTER table batting modify COLUMN SH int(3);
ALTER table batting modify COLUMN SF int(3);
ALTER table batting modify COLUMN GIDP INT(3);

~~~~

### 3) Add a Primary Key
Two options for the Primary Key:

- Composite Key: playerID, yearID, stint
- Covering Key (Index): playerID, yearID, stint, teamID


#### Choice

_Add primary key with composite key_

#### Explanation


_Alter table and add primary key_

#### SQL

~~~~sql
ALTER table batting ADD PRIMARY KEY (playerID, yearID, stint)

~~~~

#### Test

In [21]:
%sql SHOW KEYS FROM batting WHERE Key_name = 'PRIMARY' and Column_name = 'playerID'

   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
1 rows affected.


[('batting', 0, 'PRIMARY', 1, 'playerID', 'A', 20214, None, None, '', 'BTREE', '', '', 'YES', None)]

### 4) Add a foreign key on playerID between the People and Batting Tables
Note: Two people in the batting table do not exist in the people table. How should you handle this issue?

#### Explanation

_Add a foreign key reference to People table. Batting is referencing relation and People is referenced relation_

#### Change

_Alter table batting and add foreign key constraint(playerID) renferences people table (playerID)_

#### SQL

~~~~sql
ALTER table batting
ADD foreign key (playerID) references people(playerID);

~~~~

#### Tests

In [27]:
%sql Select playerID from batting where playerID not in (select playerID from people);

 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
0 rows affected.


[]

### Final CREATE Statement

To find the create statement:

- Right click on the table name in workbench
- Select 'Copy to Clipboard'
- Select 'Create Statement' 

The create statement will now be copied into your clipboard and can be pasted into the cell below. 

~~~~sql
Put your answer in this cell
~~~~

# Part D: SQL Queries

NOTE: You must use the CLEAN lahman schema provided in HW0 for the queries below to ensure your answers are consistent with the solutions. 

## Question 0
What is the highest salary in baseball history?

In [22]:
%sql mysql+pymysql://admin:Rlarudwls45!@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb

In [23]:
%sql SELECT MAX(salary) as maximum_Salary FROM salaries;

 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
1 rows affected.


[(33000000.0,)]

## Question 1
Create a Table of all players with a first name of John who were born in the United States and
played at Fordham university.

Include their first name, last name, playerID, and birth state.

<i> Hint: Use a Join between People and CollegePlaying </i>

In [94]:
%sql CREATE TABLE JOHNS SELECT nameFirst, nameLast, birthState, people.playerID as playerID, collegeplaying.schoolID as schoolID FROM people, collegeplaying WHERE people.playerID = collegeplaying.playerID and people.birthCountry = 'USA' and collegeplaying.schoolID = 'fordham' and people.nameFirst = 'john';


   mysql+pymysql://admin:***@kyungjin.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
(pymysql.err.OperationalError) (2006, "MySQL server has gone away (ConnectionResetError(10054, '현재 연결은 원격 호스트에 의해 강제로 끊겼습니다', None, 10054, None))")
[SQL: CREATE TABLE JOHNS SELECT nameFirst, nameLast, birthState, people.playerID as playerID, collegeplaying.schoolID as schoolID FROM people, collegeplaying WHERE people.playerID = collegeplaying.playerID and people.birthCountry = 'USA' and collegeplaying.schoolID = 'fordham' and people.nameFirst = 'john';]
(Background on this error at: http://sqlalche.me/e/13/e3q8)


In [24]:
%sql select * from JOHNS

 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
4 rows affected.


[('John', 'Butler', 'MA', 'butlejo01', 'fordham'),
 ('John', 'Butler', 'MA', 'butlejo01', 'fordham'),
 ('John', 'Walsh', 'PA', 'walshjo02', 'fordham'),
 ('John', 'Walsh', 'PA', 'walshjo02', 'fordham')]

## Question 2
Update all entries with full_name Columbia University to 'Columbia University!' in the Schools table. Then select the row.


In [6]:
%sql UPDATE schools SET name_full = 'Columbia University!' WHERE name_full = 'Columbia University';

 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
0 rows affected.


[]

In [25]:
%sql select * from schools where name_full = 'Columbia University!';

 * mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansbaseballdb
   mysql+pymysql://admin:***@tutorialdb1.cnaxwzfrd4me.us-east-2.rds.amazonaws.com/lahmansdb_to_clean
1 rows affected.


[('columbia', 'Columbia University!', 'New York', 'NY', 'USA')]

# Part E: CSVDataTable


## i. Conceptual Questions
The purpose of this homework is to teach you the behaviour of SQL Databases by asking you to implement functions that will model the behaviour of a real database with CSVDataTable. You will mimic a SQL Database using CSV files. 

Read through the scaffolding code provided in the CSVDataTable folder first to understand and answer the following conceptual questions.

1. Given this SQL statement:
    
        SELECT nameFirst, nameLast FROM people WHERE playerID = collied01
   
   If you run find_by_primary_key() on this statement, what are key_fields and field_list? 

    <i> Key field is "collied01" from playerID and field_list is nameFirst and nameLast </i>

2. What should be checked when you are trying to INSERT a new row into a table with a PK?

    <i>Firstly, check if primary key exists. Tyhen, check whether there is a duplicate primary key or not. If duplicate primary key exists, it should not insert a new row since PK should be unique. (That is, check the integrity constraint of primary key) </i>

3. What should be checked when you are trying to UPDATE a row in a table with a PK?

    <i> Whether primary key is valid. If PK valids, then avoid making duplicate rows. </i>

## ii. Coding


You are responsible for implementing and testing two classes in Python: CSVDataTable, BaseDataTable.
The python files and data can be found in the assignment under Courseworks. 

We have already given you **find_by_template(self, template, field_list=None, limit=None, offset=None, order_by=None)**
Use this as a jumping off point for the rest of your functions.

Methods to complete:

CSVDataTable.py
- find_by_primary_key(self, key_fields, field_list=None)
- delete_by_key(self, key_fields)
- delete_by_template(self, template)
- update_by_key(self, key_fields, new_values)
- update_by_template(self, template, new_values)
- insert(self, new_record)
CSV_table_tests.py
- You must test all methods. You will have to write these tests yourself. 
- You must test your methods on the People and Batting table.

If you do not include tests and tests outputs 50% of this section's points will be deducted at the start




## iii. Testing

Please copy the text from the output of your tests and paste it below:


============================= test session starts =============================
platform win32 -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- C:\Users\opszz\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\opszz\Desktop\Columbia University\Spring 2021\Database and System\Homework\HW1\W4111_HW1-1\W4111_HW1\4111_S2021_hw1\4111_s21_hw1_programming_CSVDataTable\tests
collecting ... collected 1 item

csv_table_tests.py::tests_people PASSED                                  [100%]
find_by_primary_key(): Known Record
{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}

find_by_primary_key(): Unknown Record
None

find_by_template(): Known Template
[{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}]

find_by_template(): Known Template
[{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}]

find_by_template(): Unknown Template
[]

delete_by_template(): Known Template
Count of deletion: 1

delete_by_template(): Unknown Template
Count of deletion: 0

delete_by_key(): Known Record
Count of deletion: 1

delete_by_key(): Known Record
Count of deletion: 1

delete_by_key(): Unknown Record
Count of deletion: 0

update_by_template(): known Template
Original Template: [{'birthDay': '15', 'birthState': 'GA'}]
Updated Template: [{'birthDay': '21', 'birthState': 'AZ'}]
Count of update: 1

update_by_template(): Unknown Template
Updated Template: []
Count of update: 0

update_by_key(): known Record
Count of update: 2

update_by_key(): Unknown Record
Count of update: 0

Insert new record - Unique Primary Key exists: 
Check if record is inserted: {'playerID': 'kjk0621', 'birthYear': '1991', 'birthMonth': '6', 'birthDay': '21'}

Insert new record - Duplicate Primary Key exists(do not insert): 
Check if record is inserted: {'playerID': 'abbotje01', 'birthYear': '1972', 'birthMonth': '8', 'birthDay': '17', 'birthCountry': 'USA', 'birthState': 'GA', 'birthCity': 'Atlanta', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'Jeff', 'nameLast': 'Abbott', 'nameGiven': 'Jeffrey William', 'weight': '190', 'height': '74', 'bats': 'R', 'throws': 'L', 'debut': '1997-06-10', 'finalGame': '2001-09-29', 'retroID': 'abboj002', 'bbrefID': 'abbotje01'}

Insert new record - Primary Key doesn't exist: 
Check if record is inserted(empty record): []

============================= test session starts =============================
platform win32 -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- C:\Users\opszz\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\opszz\Desktop\Columbia University\Spring 2021\Database and System\Homework\HW1\W4111_HW1-1\W4111_HW1\4111_S2021_hw1\4111_s21_hw1_programming_CSVDataTable\tests
collecting ... collected 1 item

csv_table_tests.py::tests_batting PASSED                                 [100%]
find_by_primary_key(): Known Record
{'playerID': 'abercda01', 'yearID': '1871', 'stint': '1', 'teamID': 'TRO', 'lgID': 'NA', 'G': '1', 'AB': '4', 'R': '0', 'H': '0', '2B': '0', '3B': '0', 'HR': '0', 'RBI': '0', 'SB': '0', 'CS': '0', 'BB': '0', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}

find_by_primary_key(): Unknown Record
None

find_by_template(): Known Template
[{'playerID': 'addybo01', 'yearID': '1871', 'stint': '1', 'teamID': 'RC1', 'lgID': 'NA', 'G': '25', 'AB': '118', 'R': '30', 'H': '32', '2B': '6', '3B': '0', 'HR': '0', 'RBI': '13', 'SB': '8', 'CS': '1', 'BB': '4', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}]

find_by_template(): Known Template
[{'playerID': 'allisdo01', 'yearID': '1871', 'stint': '1', 'teamID': 'WS3', 'lgID': 'NA', 'G': '27', 'AB': '133', 'R': '28', 'H': '44', '2B': '10', '3B': '2', 'HR': '2', 'RBI': '27', 'SB': '1', 'CS': '1', 'BB': '0', 'SO': '2', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}]

find_by_template(): Unknown Template
[]

delete_by_template(): Known Template
Count of deletion: 1

delete_by_template(): Unknown Template
Count of deletion: 0

delete_by_key(): Known Record
Count of deletion: 4

delete_by_key(): Known Record
Count of deletion: 1

delete_by_key(): Unknown Record
Count of deletion: 0

update_by_template(): known Template
Original Template: [{'yearID': '1875', 'teamID': 'CH2'}, {'yearID': '1871', 'teamID': 'CH1'}]
Updated Template: [{'yearID': '1941', 'teamID': 'CA4'}, {'yearID': '1941', 'teamID': 'CA4'}]
Count of update: 2

update_by_template(): Unknown Template
Updated Template: []
Count of update: 0

update_by_key(): known Record
Count of update: 9

update_by_key(): Unknown Record
Count of update: 0

Insert new record - Unique Primary Key exists: 
Check if record is inserted: {'playerID': 'kjk0101', 'yearID': '1991', 'teamID': 'CRI'}

Insert new record - Duplicate Primary Key exists(do not insert): 
Check if record is inserted: {'playerID': 'barre01', 'yearID': '1872', 'stint': '1', 'teamID': 'BR2', 'lgID': 'NA', 'G': '8', 'AB': '34', 'R': '6', 'H': '7', '2B': '1', '3B': '0', 'HR': '0', 'RBI': '2', 'SB': '1', 'CS': '0', 'BB': '0', 'SO': '1', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}

Insert new record - Primary Key doesn't exist: 
Check if record is inserted(empty record): []